app/controllers/likes_controller.rb

```diff
@@ -0,0 +1,23 @@
+class LikesController < ApplicationController
+  before_action :authenticate_user!, only: [:create, :destroy]
+
+  before_action :set_variables
+
+  def like
+    like = current_user.likes.new(product_id: @product.id)
+    like.save
+  end
+
+  def unlike
+    like = current_user.likes.find_by(product_id: @product.id)
+    like.destroy
+  end
+
+  private
+
+  def set_variables
+    @product = Product.find(params[:product_id])
+    @id_name = "#like-link-#{@product.id}"
```

Comment:
これはviewの中で準備できるのではないでしょうか？
わざわざインスタンス変数に入れる必要はないです。

Reply:
こちらの変数ですが、Viewではなく、JSファイルに受け渡しています。
JSファイルの内で、"#like-link-#{@product.id}"を定義したのですが、動作しなくなりました。

$("<%= @id_name %>").html('<%= escape_javascript(render("likes/like_links", product: @product  )) %>');
この記述の@id_nameに使用しているのですが、

$("<%= "#like-link-#{@product.id}" %>").html('<%= escape_javascript(render("likes/like_links", product: @product  )) %>');
としたところ、動作しなくなりました。
様々な記述を試し、機能が動作しているのが現在の記述になります。
create.jsに値を受け渡すために必要な定義だとは思うのですが、他にアプローチがあるのでしょうか？

Reply:
> $("<%= "#like-link-#{@product.id}" %>").html('<%= escape_javascript(render("likes/like_links", product: @Product )) %>');
としたところ、動作しなくなりました。
これでなぜ動作しなくなったのかを探りましょう。
<%= "#like-link-#{@product.id}" %>これが最終的にどういうコードに解釈されたのかを見ましょう

Reply:
set_variablesを使用しない場合のコードが以下のようになりました。
最初のコードを比べましたが、before_actionでインスタンス変数を定義する記述の方が記述量が短く、可読性も高いと考えています。

```suggestion
  def create
    binding.pry
    like = current_user.likes.new(product_id: params[:product_id])
    @product = Product.find(params[:product_id])
    @id_name = "#like-link-#{params[:product_id]}"
    if like.save
    else
      flash[:miss] = "いいねに失敗しました。"
    end
  end

  def destroy
    like = current_user.likes.find_by(product_id: params[:product_id])
    @product = Product.find(params[:product_id])
    @id_name = "#like-link-#{params[:product_id]}"
    if like.destroy

    else
      flash[:miss] = "いいね取り消しに失敗しました。"
    end
  end
```

before_actionで@productを再定義していた理由ですが、いいねボタンをクリック後にcerateアクションに送られるパラメータにはViewが持っていた商品情報はありません。

以下がそのcreateアクション発動時にbinding.pryで確認したparamsの中身になります。
  {"controller"=>"likes", "action"=>"create", "product_id"=>"12"}
params内には商品情報がidしかありません。

@productで定義した商品レコードがない状態では、create.jsで置き換える
render("likes/like_links", product: @Product )
に渡す@productがnillとなってしまい、動作しなくなっておりました。

この@productに情報を持たせる方法を模索しましたが、当初のコードのようにコントローラーで再定義する手段が良いと考えております。

的外れな回答でございましたら、申し訳ありません。

Reply:
すみませんが、理解できないので画面の様子をgifにおさめてもらってもよろしいですか？
いいねのcreateでhtmlを追加する意味が分からないので

Reply:
説明がいたらず、申し訳ありません。
以下がgif画像になります。
https://gyazo.com/0ac0a51c46e3a5ccd5786b34b4e1fe69

流れとしては以下の通りです。
1.商品詳細ページ表示
↓
2.いいねボタンは部分テンプレートで配置。(部分テンプレートはlike_links.htmlになります)
if文ですでにいいねされているか判別し、いいねボタンを表示。
↓
3.いいねクリック
↓
4.destroyかcreateアクションが動作
↓
5.create.jsもしくはdestroy.js内のイベント発火。
すでに表示された部分テンプレートを削除し、再度部分テンプレート読み込み
↓
6.部分テンプレート内のif文ですでにいいね済みか判別し、いいねボタンを表示
↓
7. 3.に戻る

部分テンプレート内で条件分岐させ、2種類のいいねボタンが表示されるようにしています。

参考にした記事も載せておきます。
https://qiita.com/mikan3rd/items/21d716026b51331e53a1